sw/lib/smp.c

```diff
@@ -0,0 +1,48 @@
+// Copyright 2022 ETH Zurich and University of Bologna.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Emanuele Parisi <emanuele.parisi@unibo.it>
+// Enrico Zelioli <ezelioli@iis.ee.ethz.ch>
+
+#include "smp.h"
+
+void smp_resume(void) {
+    uint32_t num_harts = *reg32(&__base_regs, CHESHIRE_NUM_INT_HARTS_REG_OFFSET);
+    // Flush cache and wake-up all sleeping cores
+    fence();
+    for (uint32_t i = 1; i < num_harts; i++) {
+        *reg32(&__base_clint, i << 2) = 0x1;
+        while (*reg32(&__base_clint, i << 2))
```

Comment:
> The smp_resume routine also waits for the interrupt to be cleared by the secondary cores before proceeding ([here](https://github.com/pulp-platform/cheshire/blob/ez/litmus/sw/lib/smp.c#L16)). This ensures that when the smp_resume returns, the IPIs have been propagated to all cores and that the cores have woke up. However, this has the downside of potentially deadlocking if another core does not wake up properly. Also, if another core has not reached the WFI loop for any reason, this will stall core 0 until then. Finally, is this really race-free?

The main possible drawback that I see here is that it might introduce a delay between waking up cores. Could the same be achieved by adding a barrier after `smp_resume` if necessary?

Reply:
Yes, we could remove the CLINT register polling and leave the synchronization up to the programmer (e.g. by adding a barrier) if needed.